// @react-shamsi/timepicker/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  TimePicker: () => TimePicker,
  defaultTimePickerTheme: () => defaultTimePickerTheme
});
module.exports = __toCommonJS(src_exports);

// src/TimePicker.tsx
var import_hooks = require("@mantine/hooks");
var import_classnames2 = __toESM(require("classnames"));
var import_persian_helpers = require("persian-helpers");
var import_react = require("react");

// src/AmPmButton.tsx
var import_classnames = __toESM(require("classnames"));
var import_jsx_runtime = require("react/jsx-runtime");
var AmPmButton = ({
  children,
  className,
  ...props
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("button", {
    type: "button",
    className: (0, import_classnames.default)("p-2 text-xs rounded-full", className),
    ...props,
    children
  });
};
var AmPmButton_default = AmPmButton;

// src/TimePicker.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var hours = [12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];
var minutes = [
  "00",
  "05",
  "10",
  "15",
  "20",
  "25",
  "30",
  "35",
  "40",
  "45",
  "50",
  "55"
];
var getDegrees = (index, items) => {
  return index * 360 / items.length;
};
var defaultTimePickerTheme = {
  backgroundColor: "#fff",
  amPmActiveBackgroundColor: "#93c5fd",
  amPmColor: "#000",
  clockBackgroundColor: "#eff6ff",
  clockLabelsColor: "#000",
  pointerBackgroundColor: "#3b82f6"
};
var TimePicker = ({
  onChange,
  theme = defaultTimePickerTheme
}) => {
  const [currentHoverIndex, setCurrentHoverIndex] = (0, import_react.useState)();
  const [debouncedCurrentHoverIndex] = (0, import_hooks.useDebouncedValue)(currentHoverIndex, 10);
  const [hour, setHour] = (0, import_react.useState)();
  const [minute, setMinute] = (0, import_react.useState)();
  const [isAm, setIsAm] = (0, import_react.useState)(true);
  const currentTimes = (0, import_react.useMemo)(() => hour ? minutes : hours, [minute, hour]);
  (0, import_react.useEffect)(() => {
    if (hour !== void 0 && minute !== void 0)
      onChange == null ? void 0 : onChange(hour, minute);
  }, [hour, minute]);
  (0, import_react.useEffect)(() => {
    if (hour) {
      setHour(
        (previousHour) => previousHour ? isAm ? previousHour - 12 : previousHour + 12 : previousHour
      );
    }
  }, [isAm]);
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)("div", {
    className: "w-full h-full relative",
    style: { backgroundColor: theme.backgroundColor },
    children: [
      /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)("div", {
        className: "w-[18rem] h-[18rem] m-8 shrink-0 relative rounded-full",
        style: { backgroundColor: theme.clockBackgroundColor },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", {
            className: "w-2 h-2 absolute top-1/2 -translate-y-1/2 left-1/2 -translate-x-1/2 rounded-full",
            style: { backgroundColor: theme.pointerBackgroundColor }
          }),
          debouncedCurrentHoverIndex !== void 0 && /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_jsx_runtime2.Fragment, {
            children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", {
              className: "absolute h-[40%] w-0.5 top-[10%] origin-bottom left-1/2 -translate-x-1/2",
              style: {
                transform: `rotate(${getDegrees(
                  debouncedCurrentHoverIndex,
                  currentTimes
                )}deg)`,
                backgroundColor: theme.pointerBackgroundColor
              }
            })
          }),
          currentTimes.map((time, index) => {
            const degrees = getDegrees(index, currentTimes);
            return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("div", {
              className: (0, import_classnames2.default)(
                "absolute text-center text-lg pointer-events-none",
                debouncedCurrentHoverIndex === index && "text-white"
              ),
              style: {
                transform: `rotate(${degrees}deg)`,
                inset: "10px"
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)("button", {
                type: "button",
                onMouseEnter: () => setCurrentHoverIndex(index),
                onMouseLeave: () => setCurrentHoverIndex(void 0),
                className: "inline-block w-8 h-8 pointer-events-auto rounded-full",
                style: {
                  transform: `rotate(-${degrees}deg)`,
                  color: debouncedCurrentHoverIndex === index ? "white" : theme.clockLabelsColor,
                  backgroundColor: debouncedCurrentHoverIndex === index ? theme.pointerBackgroundColor : "transparent"
                },
                onClick: () => {
                  if (hour)
                    setMinute(+time);
                  else
                    setHour(+time);
                },
                children: (0, import_persian_helpers.convertDigits)(time)
              })
            }, time);
          })
        ]
      }),
      /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(AmPmButton_default, {
        style: {
          backgroundColor: isAm ? theme.amPmActiveBackgroundColor : "transparent",
          color: theme.amPmColor
        },
        className: "absolute bottom-4 left-4",
        onClick: () => setIsAm(true),
        children: "\u0642\u0628\u0644 \u0627\u0632 \u0638\u0647\u0631"
      }),
      /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(AmPmButton_default, {
        style: {
          backgroundColor: isAm ? "transparent" : theme.amPmActiveBackgroundColor,
          color: theme.amPmColor
        },
        className: "absolute bottom-4 right-4",
        onClick: () => setIsAm(false),
        children: "\u0628\u0639\u062F \u0627\u0632 \u0638\u0647\u0631"
      })
    ]
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TimePicker,
  defaultTimePickerTheme
});
